### ISSUE TITLE
The ceph rbd plugin will cause the docker daemon hang after system(ubuntu 14.04) rebooted

### ISSUE THREAD

**Issue by mangobowl** [codedellemc/rexray#905]
When the first time get `rexray/rbd:0.9.2-rc2` installed on a `ubuntu 14.04` host, the plugin works OK with my ceph cluster; but after reboot the host, the docker daemon hangs, even can't be restarted by `service docker restart`. After that I install the plugin on a `ubuntu 16.04` host and do the same things , everything is OK. Is this plugin incompatible with `ubuntu 14.04`, or a bug?

```shell
Environment:

Ubuntu 14.04.5 LTS (GNU/Linux 4.4.0-81-generic x86_64)
docker 17.04.0-ce
rexray/rbd:0.9.2-rc2
```

**Comment by mangobowl** [comment]
maybe this docker logs useful.
```
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x30 pc=0x95fcdb]

goroutine 62 [running]:
panic(0x1664420, 0xc420010080)
        /usr/local/go/src/runtime/panic.go:500 +0x1a1
github.com/docker/docker/volume/store.lookupVolume(0xc42030e134, 0xa, 0xc42030e130, 0x4, 0x0, 0x0, 0x0, 0x0)
        /usr/src/docker/.gopath/src/github.com/docker/docker/volume/store/store.go:545 +0x2cb
github.com/docker/docker/volume/store.(*VolumeStore).restore.func2(0xc42030e140, 0xc4202d3c80, 0xc420270000, 0xc42030e130, 0x4, 0xc42030e134, 0xa, 0xc420257a10, 0xc420257a40)
        /usr/src/docker/.gopath/src/github.com/docker/docker/volume/store/restore.go:36 +0xf8
created by github.com/docker/docker/volume/store.(*VolumeStore).restore
        /usr/src/docker/.gopath/src/github.com/docker/docker/volume/store/restore.go:70 +0x1cb
/var/run/docker.sock is up
time="2017-06-28T17:43:30.394271959+08:00" level=info msg="libcontainerd: new containerd process, pid: 16424"
time="2017-06-28T17:43:31.404966794+08:00" level=warning msg="libcontainerd: client is out of sync, restore was called on a fully synced container (124e02fc907176f8ced85f82d9e937ddd68fdee80217147ebc6cb21df0e37c03)."
time="2017-06-28T17:43:31.407414428+08:00" level=warning msg="libcontainerd: failed to retrieve container 124e02fc907176f8ced85f82d9e937ddd68fdee80217147ebc6cb21df0e37c03 state: rpc error: code = 2 desc = containerd: container not found"
INFO[0001]                                               plugin=124e02fc907176f8ced85f82d9e937ddd68fdee80217147ebc6cb21df0e37c03
INFO[0001]                           ⌐▄Q▓▄Ç▓▄,▄_         plugin=124e02fc907176f8ced85f82d9e937ddd68fdee80217147ebc6cb21df0e37c03
INFO[0001]                          Σ▄▓▓▓▓▓▓▓▓▓▓▄π       plugin=124e02fc907176f8ced85f82d9e937ddd68fdee80217147ebc6cb21df0e37c03
INFO[0001]                        ╒▓▓▌▓▓▓▓▓▓▓▓▓▓▀▓▄▄.    plugin=124e02fc907176f8ced85f82d9e937ddd68fdee80217147ebc6cb21df0e37c03
INFO[0001]                     ,_▄▀▓▓ ▓▓ ▓▓▓▓▓▓▓▓▓▓▓█    plugin=124e02fc907176f8ced85f82d9e937ddd68fdee80217147ebc6cb21df0e37c03
INFO[0001]                    │▄▓▓ _▓▓▓▓▓▓▓▓▓┌▓▓▓▓▓█     plugin=124e02fc907176f8ced85f82d9e937ddd68fdee80217147ebc6cb21df0e37c03
INFO[0001]                   _J┤▓▓▓▓▓▓▓▓▓▓▓▓▓├█▓█▓▀Γ     plugin=124e02fc907176f8ced85f82d9e937ddd68fdee80217147ebc6cb21df0e37c03
INFO[0001]             ,▄▓▓▓▓▓▓^██▓▓▓▓▓▓▓▓▓▓▓▓▄▀▄▄▓▓Ω▄   plugin=124e02fc907176f8ced85f82d9e937ddd68fdee80217147ebc6cb21df0e37c03
INFO[0001]             F▌▓▌█ⁿⁿⁿ  ⁿ└▀ⁿ██▓▀▀▀▀▀▀▀▀▀▀▌▓▓▓▌  plugin=124e02fc907176f8ced85f82d9e937ddd68fdee80217147ebc6cb21df0e37c03
INFO[0001]              'ⁿ_  ,▄▄▄▄▄▄▄▄▄█_▄▄▄▄▄▄▄▄▄ⁿ▀~██  plugin=124e02fc907176f8ced85f82d9e937ddd68fdee80217147ebc6cb21df0e37c03
INFO[0001]                Γ  ├▓▓▓▓▓█▀ⁿ█▌▓Ω]█▓▓▓▓▓▓ ├▓    plugin=124e02fc907176f8ced85f82d9e937ddd68fdee80217147ebc6cb21df0e37c03
INFO[0001]                │  ├▓▓▓▓▓▌≡,__▄▓▓▓█▓▓▓▓▓ ╞█~   Y,┐  plugin=124e02fc907176f8ced85f82d9e937ddd68fdee80217147ebc6cb21df0e37c03
INFO[0001]                ╞  ├▓▓▓▓▓▄▄__^^▓▓▓▌▓▓▓▓▓  ▓   /▓▓▓  plugin=124e02fc907176f8ced85f82d9e937ddd68fdee80217147ebc6cb21df0e37c03
INFO[0001]                   ├▓▓▓▓▓▓▓▄▄═▄▓▓▓▓▓▓▓▓▓  π ⌐▄▓▓█║n  plugin=124e02fc907176f8ced85f82d9e937ddd68fdee80217147ebc6cb21df0e37c03
INFO[0001]                 _ ├▓▓▓▓▓▓▓▓▓~▓▓▓▓▓▓▓▓▓▓  ▄4▄▓▓▓██  plugin=124e02fc907176f8ced85f82d9e937ddd68fdee80217147ebc6cb21df0e37c03
INFO[0001]                 µ ├▓▓▓▓█▀█▓▓_▓▓███▓▓▓▓▓  ▓▓▓▓▓Ω4  plugin=124e02fc907176f8ced85f82d9e937ddd68fdee80217147ebc6cb21df0e37c03
INFO[0001]                 µ ├▓▀▀L   └ⁿ  ▀   ▀ ▓▓█w ▓▓▓▀ìⁿ  plugin=124e02fc907176f8ced85f82d9e937ddd68fdee80217147ebc6cb21df0e37c03
INFO[0001]                 ⌐ ├_                τ▀▓  Σ⌐└  plugin=124e02fc907176f8ced85f82d9e937ddd68fdee80217147ebc6cb21df0e37c03
INFO[0001]                 ~ ├▓▓  ▄  _     ╒  ┌▄▓▓  Γ    plugin=124e02fc907176f8ced85f82d9e937ddd68fdee80217147ebc6cb21df0e37c03
INFO[0001]                   ├▓▓▓▌█═┴▓▄╒▀▄_▄▌═¢▓▓▓  ╚    plugin=124e02fc907176f8ced85f82d9e937ddd68fdee80217147ebc6cb21df0e37c03
INFO[0001]                ⌠  ├▓▓▓▓▓ⁿ▄▓▓▓▓▓▓▓┐▄▓▓▓▓  └    plugin=124e02fc907176f8ced85f82d9e937ddd68fdee80217147ebc6cb21df0e37c03
INFO[0001]                Ω_.└██▓▀ⁿÇⁿ▀▀▀▀▀▀█≡▀▀▀▀▀   µ   plugin=124e02fc907176f8ced85f82d9e937ddd68fdee80217147ebc6cb21df0e37c03
INFO[0001]                ⁿ  .▄▄▓▓▓▓▄▄┌ ╖__▓_▄▄▄▄▄*Oⁿ    plugin=124e02fc907176f8ced85f82d9e937ddd68fdee80217147ebc6cb21df0e37c03
INFO[0001]                  û▌├▓█▓▓▓██ⁿ ¡▓▓▓▓▓▓▓▓█▓╪     plugin=124e02fc907176f8ced85f82d9e937ddd68fdee80217147ebc6cb21df0e37c03
INFO[0001]                  ╙Ω▀█ ▓██ⁿ    └█▀██▀▓█├█Å     plugin=124e02fc907176f8ced85f82d9e937ddd68fdee80217147ebc6cb21df0e37c03
INFO[0001]                      ⁿⁿ             ⁿ ⁿ^      plugin=124e02fc907176f8ced85f82d9e937ddd68fdee80217147ebc6cb21df0e37c03
INFO[0001] :::::::..  .,::::::    .,::      .::::::::..    :::.  .-:.     ::-.  plugin=124e02fc907176f8ced85f82d9e937ddd68fdee80217147ebc6cb21df0e37c03
INFO[0001] ;;;;'';;;; ;;;;''''    ';;;,  .,;; ;;;;'';;;;   ;;';;  ';;.   ;;;;'  plugin=124e02fc907176f8ced85f82d9e937ddd68fdee80217147ebc6cb21df0e37c03
INFO[0001]  [[[,/[[['  [[cccc       '[[,,[['   [[[,/[[['  ,[[ '[[,  '[[,[[['  plugin=124e02fc907176f8ced85f82d9e937ddd68fdee80217147ebc6cb21df0e37c03
INFO[0001]  $$$$$$c    $$""""        Y$$$Pcccc $$$$$$c   c$$$cc$$$c   c$$"  plugin=124e02fc907176f8ced85f82d9e937ddd68fdee80217147ebc6cb21df0e37c03
INFO[0001]  888b "88bo,888oo,__    oP"''"Yo,   888b "88bo,888   888,,8P"'  plugin=124e02fc907176f8ced85f82d9e937ddd68fdee80217147ebc6cb21df0e37c03
INFO[0001]  MMMM   "W" """"YUMMM,m"       "Mm, MMMM   "W" YMM   ""'mM"  plugin=124e02fc907176f8ced85f82d9e937ddd68fdee80217147ebc6cb21df0e37c03
INFO[0001]                                               plugin=124e02fc907176f8ced85f82d9e937ddd68fdee80217147ebc6cb21df0e37c03
INFO[0001] REX-Ray                                       plugin=124e02fc907176f8ced85f82d9e937ddd68fdee80217147ebc6cb21df0e37c03
INFO[0001] -------                                       plugin=124e02fc907176f8ced85f82d9e937ddd68fdee80217147ebc6cb21df0e37c03
INFO[0001] Binary: /usr/bin/rexray                       plugin=124e02fc907176f8ced85f82d9e937ddd68fdee80217147ebc6cb21df0e37c03
INFO[0001] Flavor: client+agent+controller               plugin=124e02fc907176f8ced85f82d9e937ddd68fdee80217147ebc6cb21df0e37c03
INFO[0001] SemVer: 0.9.2-rc2                             plugin=124e02fc907176f8ced85f82d9e937ddd68fdee80217147ebc6cb21df0e37c03
INFO[0001] OsArch: Linux-x86_64                          plugin=124e02fc907176f8ced85f82d9e937ddd68fdee80217147ebc6cb21df0e37c03
INFO[0001] Branch: v0.9.2-rc2                            plugin=124e02fc907176f8ced85f82d9e937ddd68fdee80217147ebc6cb21df0e37c03
INFO[0001] Commit: 1744b245121ac96b9b65ef98a80f6c9f78e4b56a  plugin=124e02fc907176f8ced85f82d9e937ddd68fdee80217147ebc6cb21df0e37c03
INFO[0001] Formed: Tue, 27 Jun 2017 14:31:45 UTC         plugin=124e02fc907176f8ced85f82d9e937ddd68fdee80217147ebc6cb21df0e37c03
INFO[0001]                                               plugin=124e02fc907176f8ced85f82d9e937ddd68fdee80217147ebc6cb21df0e37c03
INFO[0001] libStorage                                    plugin=124e02fc907176f8ced85f82d9e937ddd68fdee80217147ebc6cb21df0e37c03
INFO[0001] ----------                                    plugin=124e02fc907176f8ced85f82d9e937ddd68fdee80217147ebc6cb21df0e37c03
INFO[0001] SemVer: 0.6.2-rc1                             plugin=124e02fc907176f8ced85f82d9e937ddd68fdee80217147ebc6cb21df0e37c03
INFO[0001] OsArch: Linux-x86_64                          plugin=124e02fc907176f8ced85f82d9e937ddd68fdee80217147ebc6cb21df0e37c03
INFO[0001] Branch: v0.9.2-rc2                            plugin=124e02fc907176f8ced85f82d9e937ddd68fdee80217147ebc6cb21df0e37c03
INFO[0001] Commit: 935022d8d0d27569c270958a4b6fa6dac3a1bd69  plugin=124e02fc907176f8ced85f82d9e937ddd68fdee80217147ebc6cb21df0e37c03
INFO[0001] Formed: Tue, 27 Jun 2017 14:31:26 UTC         plugin=124e02fc907176f8ced85f82d9e937ddd68fdee80217147ebc6cb21df0e37c03
INFO[0001]                                               plugin=124e02fc907176f8ced85f82d9e937ddd68fdee80217147ebc6cb21df0e37c03
INFO[0013] time="2017-06-28T09:43:43Z" level=error msg="received signal; shutting down" signal=terminated time=1498643023546   plugin=124e02fc907176f8ced85f82d9e937ddd68fdee80217147ebc6cb21df0e37c03
time="2017-06-28T17:43:43.615629705+08:00" level=info msg="[graphdriver] using prior storage driver: aufs"
time="2017-06-28T17:43:43.621780106+08:00" level=warning msg="Unable to connect to plugin: /run/docker/plugins/124e02fc907176f8ced85f82d9e937ddd68fdee80217147ebc6cb21df0e37c03/rexray.sock/VolumeDriver.Capabilities: Post http://%2Frun%2Fdocker%2Fplugins%2F124e02fc907176f8ced85f82d9e937ddd68fdee80217147ebc6cb21df0e37c03%2Frexray.sock/VolumeDriver.Capabilities: dial unix /run/docker/plugins/124e02fc907176f8ced85f82d9e937ddd68fdee80217147ebc6cb21df0e37c03/rexray.sock: connect: no such file or directory, retrying in 1s"
time="2017-06-28T17:43:44.624663405+08:00" level=warning msg="Unable to connect to plugin: /run/docker/plugins/124e02fc907176f8ced85f82d9e937ddd68fdee80217147ebc6cb21df0e37c03/rexray.sock/VolumeDriver.Capabilities: Post http://%2Frun%2Fdocker%2Fplugins%2F124e02fc907176f8ced85f82d9e937ddd68fdee80217147ebc6cb21df0e37c03%2Frexray.sock/VolumeDriver.Capabilities: dial unix /run/docker/plugins/124e02fc907176f8ced85f82d9e937ddd68fdee80217147ebc6cb21df0e37c03/rexray.sock: connect: no such file or directory, retrying in 2s"
time="2017-06-28T17:43:46.625833258+08:00" level=warning msg="Unable to connect to plugin: /run/docker/plugins/124e02fc907176f8ced85f82d9e937ddd68fdee80217147ebc6cb21df0e37c03/rexray.sock/VolumeDriver.Capabilities: Post http://%2Frun%2Fdocker%2Fplugins%2F124e02fc907176f8ced85f82d9e937ddd68fdee80217147ebc6cb21df0e37c03%2Frexray.sock/VolumeDriver.Capabilities: dial unix /run/docker/plugins/124e02fc907176f8ced85f82d9e937ddd68fdee80217147ebc6cb21df0e37c03/rexray.sock: connect: no such file or directory, retrying in 4s"
time="2017-06-28T17:43:50.626823633+08:00" level=warning msg="Unable to connect to plugin: /run/docker/plugins/124e02fc907176f8ced85f82d9e937ddd68fdee80217147ebc6cb21df0e37c03/rexray.sock/VolumeDriver.Capabilities: Post http://%2Frun%2Fdocker%2Fplugins%2F124e02fc907176f8ced85f82d9e937ddd68fdee80217147ebc6cb21df0e37c03%2Frexray.sock/VolumeDriver.Capabilities: dial unix /run/docker/plugins/124e02fc907176f8ced85f82d9e937ddd68fdee80217147ebc6cb21df0e37c03/rexray.sock: connect: no such file or directory, retrying in 8s"
time="2017-06-28T17:43:58.627757845+08:00" level=warning msg="Volume driver rbd:latest returned an error while trying to query its capabilities, using default capabilties: Post http://%2Frun%2Fdocker%2Fplugins%2F124e02fc907176f8ced85f82d9e937ddd68fdee80217147ebc6cb21df0e37c03%2Frexray.sock/VolumeDriver.Capabilities: dial unix /run/docker/plugins/124e02fc907176f8ced85f82d9e937ddd68fdee80217147ebc6cb21df0e37c03/rexray.sock: connect: no such file or directory"
time="2017-06-28T17:43:58.628685904+08:00" level=warning msg="Unable to connect to plugin: /run/docker/plugins/124e02fc907176f8ced85f82d9e937ddd68fdee80217147ebc6cb21df0e37c03/rexray.sock/VolumeDriver.Get: Post http://%2Frun%2Fdocker%2Fplugins%2F124e02fc907176f8ced85f82d9e937ddd68fdee80217147ebc6cb21df0e37c03%2Frexray.sock/VolumeDriver.Get: dial unix /run/docker/plugins/124e02fc907176f8ced85f82d9e937ddd68fdee80217147ebc6cb21df0e37c03/rexray.sock: connect: no such file or directory, retrying in 1s"
```

**Comment by codenrhoden** [comment]
HI @mangobowl,

Well this is concerning. Let me try it out. I tested it on 16.04, and CentOS 7. I never tested 14.04, but even if I had, I wouldn't have rebooted the machine, so I probably never would have seen it.

From the log you sent, I can see that Docker starts the container with the plugin, and it looks like it is starting up fine, but then it receives a kill signal. That means Docker shut it down. But then Docker tries to connect to it's sock file, which doesn't make any sense to me. I'll look into it.

You said you had no problems with 16.04?

**Comment by clintkitson** [comment]
This sure sounds like something Docker should be aware of. If a plugin is not enabled, then I would assume Docker should not try and start the container. I am curious as to the restart policy for plugins upon Docker daemon resets and whether existing containers are able to properly locate the plugin as it launches again (if it does).

**Comment by codenrhoden** [comment]
Hi @mangobowl 

On a first pass, I was not able to recreate this. I used an Ubuntu 14.04.5 installation, stock kernel, and installed Docker 17.05.0-ce.

I got a working plugin setup on first try, like you said. I then rebooted the machine, and everything came back fine. The plugin was running, I was able to do `docker volume ls` and see my Ceph RBDs, and I was able to `docker run` containers that would mount Ceph volumes.

When I rebooted the machine, however, I did not have any existing containers running with volumes attached. Did you?

At this point, I would need more detailed logging from you. Please enable debug logging from the plugin, like so:

```
docker plugin disable -f rexray/rbd:0.9.2-rc2
docker plugin set rexray/rbd:0.9.2-rc2 REXRAY_LOGLEVEL=debug
docker plugin enable rexray/rbd:0.9.2-rc2
```

That will enable much more verbose logging, and you will see the different in the docker logs. On 14.04, those are at `/var/log/upstart/docker/log`. I think that's where you got the log you pasted before.  If you can pass those details along, that would be useful.

**Comment by mangobowl** [comment]
@codenrhoden @clintkitson  I reproduced this issue with `docker 17.04-ce`, `docker 17.05-ce` and `docker 17.06-ce` for several times, and found that `docker 17.05-ce` `docker 17.06-ce` WON'T cause docker daemon hang after reboot. I think maybe it's a compatibility issue between `docker 17.04-ce` and `ubuntu 14.04`, which be fixed after `docker 17.05-ce`. I will check it out later.

Sorry for my poor English and thanks for your patiences!
